apps/emqx_plugins/src/emqx_plugins.erl

```diff
@@ -1082,30 +1142,38 @@ read_file_fun(Path, ErrMsg, #{read_mode := ?JSON_MAP}) ->
     end.
 
 %% Directorys
+-spec plugin_dir(name_vsn()) -> string().
 plugin_dir(NameVsn) ->
-    filename:join([install_dir(), NameVsn]).
+    wrap_list_path(filename:join([install_dir(), NameVsn])).
 
+-spec plugin_config_dir(name_vsn()) -> string().
 plugin_config_dir(NameVsn) ->
```

Comment:
Not related to this PR, but something we might have missed before for plugins in general:
we have to think about how to support upgrades if name-vsn is in the path to plugin data directory.

We should probably put plugins data in EMQX's data directory, and not include `vsn` part in the path,
so that when a plugin upgrades from for example `0.1.0` to `0.1.1`, it can still load old config.

In the plugin example, we can probably put a version number in the config schema, so to encourage developers to think about compatibility.
e.g. for configs, there is a `version` field, so the plugin code can perform upgrade-on-read.

Reply:
But this cannot handle the situation where the schema is changed when the plug-in version is upgraded.
In other words, schema, configuration, and plugin version should be strongly related.
But in order to preserve the old configuration during the plugin upgrade process. We can store it in the data directory.
nit: Upgrading means deleting the old version of the plugin and installing the new version.

Reply:
the plugin should export a upgrade_config callback which will be called by plugin framework after loading the config, but before checking the config against the schema.